### DISCUSSION TITLE
Improving geo-search: bounding box or geometry sort

### DISCUSSION THREAD

**Discussion by curquiza** [meilisearch/product#431] [Feedback & Feature Proposal]
Because there is still the GitHub bug, I cannot transfer the discussion, this discussion is following this one: https://github.com/meilisearch/meilisearch/discussions/2302

By @questionablequestion

> Hi all,
>
> Currently there is a Geo Search functionality included but it seems very much limited, at least for my case. I would need a possibility to filter records based on wether they are in the bounding box or geometry, which isn't possible now. I see that was planned on the roadmap (https://roadmap.meilisearch.com/c/33-geo-search) but I can't find any more info since July 2021 (comment from the roadmap item).
>
> Is there any info on when can we expect further improvements on Geo Search? I really like the simplicity of working with Meilisearch but I'm afraid if there's nothing coming soon, I would have to settle with typesense or something similar.
>
> Thanks!

**Comment by irevoire** [comment]
Hey, @questionablequestion, I was wondering;
1. Do you want to send a bounding box at **search time** and get the documents that are in this box?
2. Or do you want to **index** documents containing bounding boxes, and then, **at search time**, see in which bounding boxes is a point.
3. Both

**Comment by questionablequestion** [comment]
Hi @irevoire, 1. would be sufficient for my use case. Example would be a user moving/zooming a map, and filtering points with coordinates within map boundary. While I could do that already with the current filter, it would be good to have specific (and maybe more efficient) filter for that and my main concern and deal breaker currently is filtering within geometry.

**Reply by irevoire** [reply]
Glad to hear! I think that's the easier to implements and also the most demanded!

Do you have an idea of the syntax you would like to use to describe a geometry?

I'm not a fan of a "function" looking like that;
```
_geoPolygon((lat1, lng1), (lat2, lng2), (lat3, lng3))
```

But I don't see a lot of other options 😒

**Reply by questionablequestion** [reply]
That would be fine, or if it's possible, an array of coordinates (as per geojson definition) would also be fine. Do you think that MultiPolygon would be easy to implement as well - if so then array with array of coordinates would make more sense if it's possible? 

I'm asking for too much, honestly, I would be happy with whatever syntax as long as it's works 😄

**Reply by irevoire** [reply]
Ah yes, in my example, I used `(` everywhere, but anything could be used really.
And actually, I imagined the `_geoPolygon` thingy to be a « MultiPolygon » with as many points as you want!
```
1: _geoPolygon([[lat1, lng1], [lat2, lng2], [lat3, lng3], ...])
-> the function takes a variable size array of array of 2 floats

2: _geoPolygon([lat1, lng1], [lat2, lng2], [lat3, lng3], ...)
-> The function takes a variable number of arrays of 2 floats

3:_geoPolygon((lat1, lng1), (lat2, lng2), (lat3, lng3), ...)
-> The function takes a variable number of tuples of 2 floats
```

So here, from what I understand, the most « expected » syntax would be the first one because it's close to the geojson definition :grin: 
(let me know if I'm wrong or if you want to add something, I don't know much about the usage of the geosearch feature. Even though I'll probably be the one implementing this feature in meilisearch)

**Reply by questionablequestion** [reply]
You are spot on with the first one! I will be here as much as I can to help you with whatever I can 😃

**Reply by questionablequestion** [reply]
Hey @irevoire , do you have any rough estimate on when could we expect this feature? Thanks

**Reply by irevoire** [reply]
Hey, no, sorry, I don't think it's being prioritized at all right now, but I would love to work on it, so maybe we could do something about it in a few months when I have some time :pensive: 

Maybe @gmourier will have a better answer though

**Reply by gmourier** [reply]
Hello @questionablequestion 👋

@irevoire is right, this is something that is not prioritized at this time but will definitely be. I'd like to see it added by the end of the year though but we have more impactful topics before that like multi-index search or an improvement in the way facets can be used

**Reply by questionablequestion** [reply]
Thanks for the quick answer! I totally understand and I can't wait for it so I can eventually transfer back to meilisearch 😃  In the meantime, cheers and good luck!

**Reply by farfromrefug** [reply]
Just to mention i also interested by this. I am investigating the use of milli as a geocoding service.

**Comment by curquiza** [comment]
Hello @gmourier! A really involved contributor would like to implement this, if this is something we want to prioritize, and if the product work is not too much to do 😄 
Following: https://github.com/meilisearch/meilisearch/issues/2757

**Reply by gmourier** [reply]
Sure, @curquiza!

On the usage side, here is what we imagined some time ago without being able to implement it due to other 🔥 subjects.

Filter documents by a given geo bounding box using the built-in filter `_geoBoundingBox([{lat, lng}, {lat, lng}])`.

`[{lat, lng} <- Top left corner, {lat, lng} <- Bottom right corner]`

We could also allow the capacity to write it with a fully described bounding box with 4 coordinates. (Top left corner, top right corner, bottom right corner, bottom left corner; clockwise drawing). `_geoBoundingBox([{lat, lng}, {lat, lng}, {lat, lng}, {lat, lng}])`.

Knowing that we could introduce polygon support in the future with something like that `_geoPolygon([{lat, lng}, ...])`

WDYT? @farfromrefug, @questionablequestion, @mlemesle, @irevoire, @curquiza.

Edit: Re-reading the discussion; It seems to be a convenient solution, but it has been a while. If we can have a double-check on that, it's super appreciated.

Thanks!

**Reply by mlemesle** [reply]
Hey @gmourier ! Thanks for answering so quickly.
So if I understand correctly, the query with 2 or 4 points should result in the same shape which is a box?
Allowing 4 points won't enforce more checks during the parsing ?
What I mean is that with 2 points you have a box, with 4 points you can have something that isn't a box at all.
So either we'll need to add checks to ensure that the 4 points actually describe a box (where 2 points are enough).
Either the four points describe something else that isn't a box and then it basically is geoPolygon with 4 points.

**Reply by gmourier** [reply]
That's a good point, @mlemesle. I agree with you; it's a polygon in this case.

I have the impression that accepting only two points to draw a bounding box will meet most people's needs, and that's more than enough to bring a valuable addition to Meilisearch.

If there is an interest in accepting 4 points and verifying that it is a bounding box, we can consider it later, depending on the feedback.

**Reply by irevoire** [reply]
Second point! Doing these extra checks with floats and expecting them to fall exactly where they’re supposed to be is probably a bad idea since floats are so bad at being precise 😬 
I think it would just be a bad API that could be misunderstood with the `_geoPolygon` function that we could add later 👀

**Reply by farfromrefug** [reply]
`_geoBoundingBox` only is fine with me

**Reply by gmourier** [reply]
Thanks @farfromrefug

**Reply by questionablequestion** [reply]
I agree with having 2 points to define bounding box will be fine for most cases, and would also love to see polygon search implemented later on for more complex cases.

**Reply by gmourier** [reply]
Thanks @questionablequestion

**Reply by curquiza** [reply]
Issue created, ready to be implemented 🔥 
https://github.com/meilisearch/meilisearch/issues/2761

**Comment by seifrached** [comment]
hello, for this moment I use as a solution this filter:

 `_geo.lat '${sw_lat}' TO '${ne_lat}' AND _geo.lng '${sw_lng}' TO '${ne_lng}'`

**Comment by gmourier** [comment]
Hello, everyone following this discussion 👋 

We have just released the first RC (release candidate) of Meilisearch containing the **bounding box** support!

You can test it by using
-  the [release assets](https://github.com/meilisearch/meilisearch/releases)
- the [Meilisearch Docker image](https://hub.docker.com/repository/docker/getmeili/meilisearch/general)

```bash
docker run -it --rm -p 7700:7700 -v $(pwd)/meili_data:/meili_data getmeili/meilisearch:v1.1.0-rc.0
```

You are welcome to communicate any feedback about this new implementation in this discussion.

If you encounter any bugs, please report them [here](https://github.com/meilisearch/meilisearch/issues/new?assignees=&labels=&template=bug_report.md&title=).
Thanks in advance for your help and your involvement in Meilisearch ❤️ 

🎉 Official and stable release containing this change will be available on 3rd April 2023

⚠️ RC (release candidates) are not recommended for production

**Comment by gmourier** [comment]
Hey folks 👋

[v1.1](https://github.com/meilisearch/meilisearch/releases/tag/v1.1.0) has been released and features the support of a geo-bounding box filter ✨

📚 https://docs.meilisearch.com/learn/advanced/geosearch.html#geoboundingbox
